jfr/src/main/java/io/netty/jfr/JfrChannelHandler.java

```diff
@@ -0,0 +1,417 @@
+    }
+
+    @Override
+    public void bind(ChannelHandlerContext ctx, SocketAddress localAddress, ChannelPromise promise) {
+        final BindEvent bindEvent = new BindEvent();
+        bindEvent.begin();
+        ctx.bind(localAddress, promise);
+        bindEvent.end();
+        if (bindEvent.shouldCommit()) {
+            bindEvent.setChannel(ctx.channel());
+            bindEvent.setLocalAddress(localAddress);
+            bindEvent.commit();
+        }
+    }
+
+    @Override
+    public void connect(
+            ChannelHandlerContext ctx,
+            SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
+        final ConnectEvent connectEvent = new ConnectEvent();
+        connectEvent.begin();
+        ctx.connect(remoteAddress, localAddress, promise);
+        connectEvent.end();
+        if (connectEvent.shouldCommit()) {
+            connectEvent.setChannel(ctx.channel());
+            connectEvent.setRemoteAddress(remoteAddress);
+            connectEvent.setLocalAddress(localAddress);
+            connectEvent.commit();
+        }
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) {
+        final DisconnectEvent disconnectEvent = new DisconnectEvent();
+        disconnectEvent.begin();
+        ctx.disconnect(promise);
+        disconnectEvent.end();
+        if (disconnectEvent.shouldCommit()) {
+            disconnectEvent.setChannel(ctx.channel());
+            disconnectEvent.commit();
+        }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) {
+        final CloseEvent closeEvent = new CloseEvent();
+        closeEvent.begin();
+        ctx.close(promise);
+        closeEvent.end();
+        if (closeEvent.shouldCommit()) {
+            closeEvent.setChannel(ctx.channel());
+            closeEvent.commit();
+        }
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) {
+        final DeregisterEvent deregisterEvent = new DeregisterEvent();
+        deregisterEvent.begin();
+        ctx.deregister(promise);
+        deregisterEvent.end();
+        if (deregisterEvent.shouldCommit()) {
+            deregisterEvent.setChannel(ctx.channel());
+            deregisterEvent.commit();
+        }
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) {
+        final ChannelReadCompleteEvent channelReadCompleteEvent = new ChannelReadCompleteEvent();
+        channelReadCompleteEvent.begin();
+        ctx.fireChannelReadComplete();
+        channelReadCompleteEvent.end();
+        if (channelReadCompleteEvent.shouldCommit()) {
+            channelReadCompleteEvent.setChannel(ctx.channel());
+            channelReadCompleteEvent.commit();
+        }
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        final ChannelReadEvent channelReadEvent = new ChannelReadEvent();
```

Comment:
I've spoken to @egahlin and he says that the allocations will most likely be optimized and removed, right?
But I could try this approach and do some comparisons.

Reply:
Generally speaking I think it is a good idea to not allocate if we can ... Just ensure the handler is not sharable anymore in this case as it has state.

Reply:
I have just looked at the `jdk.jfr.events.SocketReadEvent` in the OpenJDK 11 and they are using the same trick.

The IO threading model is different, but I think the `ThreadLocal` can be applied.

Reply:
> I've spoken to @egahlin and he says that the allocations will most likely be optimized and removed, right?
> But I could try this approach and do some comparisons.

Oh! I know that JFR trigger a special class loading code, but I haven't dived too much how it works (how the event allocation is being optimized away during class-loading? or via the jit?).

Reply:
If the JIT is able to eliminate the heap allocation, it is (at least in theory) cheaper than a TL.

https://youtu.be/plYESjZ12hM?t=1126

It also works very nicely if the event is disabled.

We are trying to get rid TLs in the JDK because of Loom. We reworked SocketReadEvent so it no longer uses TLS. I think it is best measure, but be aware of the future.

Reply:
@egahlin Thank you for this insight ! And for the presentation, it expalins very well how this works !